File: ctsa-capstone-microservices/results-suggestions-service/src/main/java/io/ctsa/resultssuggestionsservice/suggestion/recommendation/PupilSuggestion.java
package io.ctsa.resultssuggestionsservice.suggestion.recommendation;

import io.ctsa.resultssuggestionsservice.model.*;
import io.ctsa.resultssuggestionsservice.repository.EntranceExamResultCentroidRepository;
import io.ctsa.resultssuggestionsservice.repository.HighSchoolTopResultCentroidRepository;
import io.ctsa.resultssuggestionsservice.repository.MajorCentroidRepository;
import io.ctsa.resultssuggestionsservice.repository.MajorRepository;
import org.springframework.stereotype.Component;

import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import static java.lang.Math.pow;
import static java.lang.Math.sqrt;

@Component
public class PupilSuggestion {

    private final MajorCentroidRepository majorCentroidRepository;

    private final HighSchoolTopResultCentroidRepository highSchoolTopResultCentroidRepository;

    private final MajorRepository majorRepository;

    private final EntranceExamResultCentroidRepository entranceExamResultCentroidRepository;

    public PupilSuggestion(MajorCentroidRepository majorCentroidRepository,
                           HighSchoolTopResultCentroidRepository highSchoolTopResultCentroidRepository,
                           MajorRepository majorRepository,
                           EntranceExamResultCentroidRepository entranceExamResultCentroidRepository) {
        this.majorCentroidRepository = majorCentroidRepository;
        this.highSchoolTopResultCentroidRepository = highSchoolTopResultCentroidRepository;
        this.majorRepository = majorRepository;
        this.entranceExamResultCentroidRepository = entranceExamResultCentroidRepository;
    }

    public List<MajorSuggestion> suggest(SuggestedMajor input) {

        return majorCentroidRepository.findAll()
                                      .stream()
                                      .map(majorCentroid -> {
                                          MajorSuggestion suggestion = new MajorSuggestion();
                                          suggestion.setMajor(majorRepository.findById(majorCentroid.getMajorId())
                                                                             .orElse(null));
                                          suggestion.setDistance(score(input, majorCentroid));
                                          return suggestion;
                                      })
                                      .sorted(Comparator.comparingDouble(MajorSuggestion::getDistance))
                                      .collect(Collectors.toList());
    }

    private double score(SuggestedMajor input, MajorCentroid centroid) {
        List<HighSchoolTopResultCentroid> highSchoolTopResultCentroids = highSchoolTopResultCentroidRepository
                .findByMajorCentroidId(centroid.getMajorId());

        double squaredDistance = pow(input.getCharacteristic() - centroid.getCharacteristic(), 2)
                + pow(input.getHighSchoolAverage() - centroid.getHighSchoolAverage(), 2)
                + scoreTopSubjects(input.getHighSchoolTopInputs(), highSchoolTopResultCentroids);

        if (input.getEntranceExamInput() != null) {
            squaredDistance += scoreEntranceExam(input.getEntranceExamInput(),
                                                 Objects.requireNonNull(
                                                         entranceExamResultCentroidRepository
                                                                 .findById(centroid.getEntranceExamResultCentroidId())
                                                                 .orElse(null)));
        }

        return sqrt(squaredDistance);
    }

    private double scoreTopSubjects(List<HighSchoolTopInput> inputs, List<HighSchoolTopResultCentroid> centroids) {
        double squaredDistance = 0.0;

        for (HighSchoolTopInput input : inputs) {
            for (HighSchoolTopResultCentroid centroid : centroids) {
                if (input.getSubjectId() == centroid.getSubjectId()) {
                    squaredDistance += pow(input.getMark() * input.getWeight() - centroid.getMark() * centroid.getWeight(), 2);
                }
            }
        }

        return squaredDistance;
    }

    private double scoreEntranceExam(EntranceExamInput input, EntranceExamResultCentroid centroid) {
        double squaredDistance = pow(input.getMath() - centroid.getMath(), 2)
                + pow(input.getLiterature() - centroid.getLiterature(), 2)
                + pow(input.getEnglish() - centroid.getEnglish(), 2);

        squaredDistance += input.getPhysics() != null && centroid.getPhysics() != null ?
                           pow(input.getPhysics() - centroid.getPhysics(), 2) : 0;

        squaredDistance += input.getChemistry() != null && centroid.getChemistry() != null ?
                           pow(input.getChemistry() - centroid.getChemistry(), 2) : 0;

        squaredDistance += input.getBiology() != null && centroid.getBiology() != null ?
                           pow(input.getBiology() - centroid.getBiology(), 2) : 0;

        squaredDistance += input.getHistory() != null && centroid.getHistory() != null ?
                           pow(input.getHistory() - centroid.getHistory(), 2) : 0;

        squaredDistance += input.getGeography() != null && centroid.getGeography() != null ?
                           pow(input.getGeography() - centroid.getGeography(), 2) : 0;

        return squaredDistance;
    }
}

File: ctsa-capstone-microservices/results-suggestions-service/src/main/java/io/ctsa/resultssuggestionsservice/suggestion/recommendation/PupilSuggestion.kt
package io.ctsa.resultssuggestionsservice.suggestion.recommendation

import io.ctsa.resultssuggestionsservice.model.centroid.EntranceExamCentroid
import io.ctsa.resultssuggestionsservice.model.centroid.HighSchoolTopCentroid
import io.ctsa.resultssuggestionsservice.model.centroid.MajorCentroid
import io.ctsa.resultssuggestionsservice.model.input.EntranceExamInput
import io.ctsa.resultssuggestionsservice.model.input.HighSchoolTopInput
import io.ctsa.resultssuggestionsservice.model.input.SuggestedMajorInput
import io.ctsa.resultssuggestionsservice.repository.centroid.EntranceExamCentroidRepository
import io.ctsa.resultssuggestionsservice.repository.centroid.HighSchoolTopResultCentroidRepository
import io.ctsa.resultssuggestionsservice.repository.centroid.MajorCentroidRepository
import io.ctsa.resultssuggestionsservice.repository.common.MajorRepository
import org.springframework.stereotype.Component
import java.util.Comparator
import java.util.Objects
import java.util.stream.Collectors

import java.lang.Math.pow
import java.lang.Math.sqrt

@Component
class PupilSuggestion(private val majorCentroidRepository: MajorCentroidRepository,
                      private val highSchoolTopResultCentroidRepository: HighSchoolTopResultCentroidRepository,
                      private val majorRepository: MajorRepository,
                      private val entranceExamCentroidRepository: EntranceExamCentroidRepository) {

    fun suggest(input: SuggestedMajorInput): List<MajorSuggestion> {
        return majorCentroidRepository.findAll()
                .asSequence()
                .map { majorCentroid ->
                    val suggestion = MajorSuggestion()

                    suggestion.major = majorRepository
                            .findById(majorCentroid.majorId)
                            .orElse(null)
                    suggestion.distance = score(input, majorCentroid)

                    suggestion
                }
                .sortedBy { it.distance }
                .toList()
    }

    private fun score(input: SuggestedMajorInput, centroid: MajorCentroid): Double {
        val highSchoolTopCentroids = highSchoolTopResultCentroidRepository
                .findByMajorCentroidId(centroid.id)

        var squaredDistance = (pow(input.characteristic - centroid.characteristic, 2.0)
                               + pow(input.highSchoolAverage!! - centroid.highSchoolAverage, 2.0)
                               + scoreTopSubjects(input.highSchoolTopInputs, highSchoolTopCentroids))

        if (input.entranceExamInput != null) {
            squaredDistance += scoreEntranceExam(input.entranceExamInput,
                                                 entranceExamCentroidRepository
                                                         .findById(centroid.entranceExamCentroidId)
                                                         .orElse(null))
        }

        return sqrt(squaredDistance)
    }

    private fun scoreTopSubjects(inputs: Collection<HighSchoolTopInput>, centroids: Collection<HighSchoolTopCentroid>): Double {
        var squaredDistance = 0.0

        inputs.forEach { input ->
            centroids.forEach { centroid ->
                if (input.subjectId == centroid.subjectId) {
                    squaredDistance += pow(input.mark * input.weight - centroid.mark * centroid.weight, 2.0)
                }
            }
        }

        return squaredDistance
    }

    private fun scoreEntranceExam(input: EntranceExamInput, centroid: EntranceExamCentroid): Double {
        var squaredDistance = 0.0

        input.entranceExamInputDetails.forEach { inputDetail ->
            centroid.entranceExamCentroidDetails.forEach { centroidDetail ->
                if (inputDetail.entranceExamSubjectId == centroidDetail.entranceExamSubjectId) {
                    squaredDistance += pow(inputDetail.mark - centroidDetail.mark * centroidDetail.weight, 2.0)
                }
            }
        }

        return squaredDistance
    }
}
